### PR TITLE
Migrated legacy multiplayer per entity analytics over to O3DE Multiplayer Gem

### PR DESCRIPTION
**Pull request by AMZN-Olex**
Added a new ImGui menu under Multiplayer: "Multiplayer Entity Stats"

image

Added a debug overlay on per entity basis:

image

It can be configured by the following cvars:
- net_DebugEntities_ShowBandwidth (false by default. if true, imgui menu is not necessary)
- net_DebugEntities_ShowAboveKbps
- net_DebugEntities_WarnAboveKbps
- net_DebugEntities_WarningColor
- net_DebugEntities_BelowWarningColor

### CODE REVIEW

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    {
+        ReportBytes(m_aggregateBytes);
+        m_aggregateBytes = 0;
+    }
+
+    float MultiplayerDebugByteReporter::GetAverageBytes() const
+    {
+        if (m_count == 0)
+        {
+            AZ_Warning("MultiplayerDebugByteReporter", m_totalBytes == 0, "Attempted to average bytes with a zero count.");
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:46]: This seems like a reasonable edge case, I'm not sure a warning is necessary here.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        m_componentDirtyBytes.Combine(other.m_componentDirtyBytes);
+    }
+
+    void MultiplayerDebugEntityReporter::ReportField(AZ::u32 index, const char* componentName,
+        const char* fieldName, size_t byteSize)
+    {
+        if (m_currentComponentReport == nullptr)
+        {
+            std::stringstream component;
+            component << "[" << std::setw(2) << std::setfill('0') << static_cast<int>(index) << "]" << " " << componentName;
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:161]: NP: Might want to consider aznumeric_cast here due to standards

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    void MultiplayerDebugEntityReporter::Reset()
+    {
+        MultiplayerDebugByteReporter::Reset();
+
+        m_componentReports.clear();
+    }
+
+    AZStd::map<AZStd::string, MultiplayerDebugComponentReporter>& MultiplayerDebugEntityReporter::GetComponentReports()
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:199]: Just to check, is it possible for this to be const/const ref or is the current form intentional?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        m_fieldReports[fieldName].ReportBytes(byteSize);
+    }
+
+    void MultiplayerDebugComponentReporter::ReportFragmentEnd()
+    {
+        MultiplayerDebugByteReporter::ReportAggregateBytes();
+        m_componentDirtyBytes.ReportAggregateBytes();
+    }
+
+    AZStd::vector<MultiplayerDebugComponentReporter::Report> MultiplayerDebugComponentReporter::GetFieldReports()
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:125]: Just to check, this is intentionally not const due to the ByteReporter pointer in the Report type?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+        m_componentDirtyBytes.Combine(other.m_componentDirtyBytes);
+    }
+
+    void MultiplayerDebugEntityReporter::ReportField(AZ::u32 index, const char* componentName,
+        const char* fieldName, size_t byteSize)
+    {
+        if (m_currentComponentReport == nullptr)
+        {
+            std::stringstream component;
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:160]: Is std::stringstream strictly necessary here due to the formatting calls?

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp (hunk 3 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:199]: I tried to make it a const but, alas, the code underneath won't work with const here, as it reset states under the hood during this call.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp (hunk 4 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:125]: Yeah, it's the intentional design here. Same as GetComponentReports()

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp (hunk 5 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:160]: It is needed for the fancy formatting that this code was written for. Specifically, the use of std::setfill.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp (hunk 1 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:46]: That is fair, removing.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp (hunk 2 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugByteReporter.cpp:161]: Sounds good, will do.

**Review by AMZN-puvvadar — APPROVED** [review]

**Review by AMZN-Gene — APPROVED** [review]

File: Gems/Multiplayer/Code/CMakeLists.txt [hunk 6]
```diff
@@ -79,7 +79,7 @@ ly_add_target(
 
 # The  "Multiplayer" target is used by clients and servers, Debug is used only on clients.
 ly_create_alias(NAME Multiplayer.Clients NAMESPACE Gem TARGETS Gem::Multiplayer Gem::Multiplayer.Debug)
-ly_create_alias(NAME Multiplayer.Servers NAMESPACE Gem TARGETS Gem::Multiplayer)
+ly_create_alias(NAME Multiplayer.Servers NAMESPACE Gem TARGETS Gem::Multiplayer Gem::Multiplayer.Debug)
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/CMakeLists.txt:82]: I keep doing this locally as well, we should probably just do this until the server target actually becomes headless.

File: Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugPerEntityReporter.cpp [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+#include <imgui/imgui.h>
+#endif
+
+AZ_CVAR(float, net_DebugEntities_ShowAboveKbps, 1.f, nullptr, AZ::ConsoleFunctorFlags::Null,
+    "Prints bandwidth on network entities with higher kpbs than this value");
+
+AZ_CVAR(float, net_DebugEntities_WarnAboveKbps, 10.f, nullptr, AZ::ConsoleFunctorFlags::Null,
+    "Prints bandwidth on network entities with higher kpbs than this value");
+
+AZ_CVAR(AZ::Color, net_DebugEntities_WarningColor, AZ::Colors::Red, nullptr, AZ::ConsoleFunctorFlags::Null,
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugPerEntityReporter.cpp:26]: Comments are off for the two colours... copy/paste error

Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugPerEntityReporter.cpp:26]: Good catch, i'll take care of it in a followup.

**Pull request merged**